tests/lib/nested.sh

```diff
@@ -0,0 +1,169 @@
+#!/bin/bash
+
+WORK_DIR="${WORK_DIR:-/tmp/work-dir}"
+SSH_PORT=${SSH_PORT:-8022}
+MON_PORT=${MON_PORT:-8888}
+IMAGE_FILE="${WORK_DIR}/ubuntu-core-20.img"
+
+execute_remote(){
+    sshpass -p ubuntu ssh -p "$SSH_PORT" -o ConnectTimeout=10 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no ubuntu@localhost "$*"
+}
+
+wait_for_service() {
+    local service_name="$1"
+    local state="${2:-active}"
+    for i in $(seq 300); do
+        if systemctl show -p ActiveState "$service_name" | grep -q "ActiveState=$state"; then
+            return
+        fi
+        # show debug output every 1min
+        if [ "$i" -gt 0 ] && [ $(( i % 60 )) = 0 ]; then
+            systemctl status "$service_name" || true;
+        fi
+        sleep 1;
+    done
+
+    echo "service $service_name did not start"
+    exit 1
+}
+
+wait_for_ssh(){
+    retry=400
+    wait=1
+    while ! execute_remote true; do
+        retry=$(( retry - 1 ))
+        if [ $retry -le 0 ]; then
+            echo "Timed out waiting for ssh. Aborting!"
+            return 1
+        fi
+        sleep "$wait"
+    done
+}
+
+cleanup_nested_core_vm(){
+    # stop the VM if it is running
+    systemctl stop nested-vm-*
+
+    # remove the swtpm
+    # TODO: we could just remove/reset the swtpm instead of removing the snap 
+    # wholesale
+    snap remove swtpm-mvo
+
+    # delete the image file
+    rm -rf "${IMAGE_FILE}"
+}
+
+start_nested_core_vm_unit(){
+    # copy the image file to create a new one to use
+    # TODO: maybe create a snapshot qcow image instead?
+    mkdir -p "${WORK_DIR}"
+    cp /home/core-initrd/pc.img "${IMAGE_FILE}"
+
+    # use only 2G of RAM for qemu-nested
+    if [ "${SPREAD_BACKEND}" = "google-nested" ]; then
+        PARAM_MEM="-m 4096"
+        PARAM_SMP="-smp 2"
+    elif [ "${SPREAD_BACKEND}" = "qemu-nested" ]; then
+        PARAM_MEM="-m 2048"
+        PARAM_SMP="-smp 1"
+    else
+        echo "unknown spread backend ${SPREAD_BACKEND}"
+        exit 1
+    fi
+
+    PARAM_DISPLAY="-nographic"
+    PARAM_NETWORK="-net nic,model=virtio -net user,hostfwd=tcp::${SSH_PORT}-:22"
+    # TODO: do we need monitor port still?
+    PARAM_MONITOR="-monitor tcp:127.0.0.1:${MON_PORT},server,nowait"
+    PARAM_RANDOM="-object rng-random,id=rng0,filename=/dev/urandom -device virtio-rng-pci,rng=rng0"
+    PARAM_CPU=""
+    PARAM_TRACE="-d cpu_reset"
+    PARAM_LOG="-D ${WORK_DIR}/qemu.log"
+    PARAM_SERIAL="-serial file:${WORK_DIR}/serial.log"
+    PARAM_TPM=""
+
+    ATTR_KVM=""
+    if [ "$ENABLE_KVM" = "true" ]; then
+        ATTR_KVM=",accel=kvm"
+        # CPU can be defined just when kvm is enabled
+        PARAM_CPU="-cpu host"
+        # Increase the number of cpus used once the issue related to kvm and ovmf is fixed
+        # https://bugs.launchpad.net/ubuntu/+source/kvm/+bug/1872803
```

Comment:
Above was fixed in june, so can we try again with more CPU cores nested in Google cloud?

Reply:
sure go for it, but every time we try to use faster / better VM's in GCE it doesn't work for us in weird ways, so YMMV